[RFC] stm32: clock_control: Enable device clock source selection

erwango opened the pull request:
**EDIT 08'22:** 
This PR was some time ago and is now largely in use. But I can now see that "alternative/alternate/optional" wording is confusing and "domain" clock would have better a better choice. If you read these lines, please keep this in mind. Follow up change: https://github.com/zephyrproject-rtos/zephyr/pull/48679

## TL;DR ##
STM32 clock_control driver update to provide a way to configure alternate clocks on peripherals that support it:
```
    clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
             <&rcc STM32_SRC_PLL3_P SPI123_SEL(2)>;   <- Newly added alternate clock (PLL3_P output here)
```
This alternate clock source could be selected in peripheral driver thanks to new `clock_control_configure()` API:
```
	if (IS_ENABLED(STM32_SPI_OPT_CLOCK_SUPPORT) && (cfg->pclk_len > 1)) {
		if (clock_control_configure(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
					    (clock_control_subsys_t) &cfg->pclken[1],
					    NULL) != 0) {
			LOG_ERR("Could not select alternate SPI source clock");
			return -EIO;
		}
	}
```

Follow up of https://github.com/zephyrproject-rtos/zephyr/pull/42097 which was limited to H7.

## Goal
Aim of this work is to enable the use of an optional clock in a device configuration.
This is particularly required in case of STM32H7 based platform on which device clock configuration is usually based on 2 different clocks (Fixes #41650).
- A register clock, required for device configuration, usually coming from APB/AHB bus.
- A kernel clock, which will be used to clock the hardware block (and hence is the one driving the device frequency), typically derived from an alternate clck source (PLL output, fixed clock, ..)

On other STM32 series, provide the possibility to provide an alternate clock to a peripheral. For instance, configuring LSE as the LPUART or LPTIM clock, or HSI48 as the USB clock.

## Update of clock_control API (03/14):
 
Reworked to make use of new `clock_control_configure()` API, cf https://github.com/zephyrproject-rtos/zephyr/pull/43790

## Implementation options: 

### Alternate clocks bindings
Additional/Alternate clocks could be fixed clocks (HSI, HSE, ..), PLL outputs (PLLX_P/Q/R), sysclock, ....
For all these clocks, I've decided to keep existing RCC device driver as the clock controller. As a consequence, the dt binding used to select a specific clock looks to:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
		 <&rcc STM32_SRC_PLL3_P SPI123_SEL(2)>;   <- Newly added alternate clock (PLL3_P output here)
```
Another way to support this would have been to set up each potential clock source as a clock controler on its own and use the following binding:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,  <- Existing, bus driven, "register" clock
		 <&pll3 pll_p_output SPI123_SEL(2)>;      <- Discarded binding option
```
I've discarded this second option because, to be consistent with the binding, this second option would require to instantiate each potential source clock as a device driver, each one implementing clock_control API. This would have had negative impact on code footprint, which could have been considered ok for STM32H7 series but problematic on "smaller" STM32 series.

### Use of clock-names
In #42097, I tried to comply with H7 reference manuals (`clock-names = "reg", "kernel";`). But extending to other series this was hard to find a single nomenclature matching all series. Besides, usage of `clock-names` didn't show any benefit of code simplification.
As a consequence, in this PR, I'm setting the use of `clock-names` property to optional. It is consumer driver responsibility to access and correctly use clock source cell.

### Clock mux bindings
Complete solution requires to find a way to configure clock mux (as per_ck).

First option is to add a dedicated node descibing  the clock mux. Its only property is to configure its clock input
```
&perck {
	clocks = <&rcc STM32_SRC_HSI_KER CKPER_SEL(0)>;
	status = "okay";
};
```
Works fine, main drawback is the cost (~220 bytes of flash)

A much simpler alternative was to add clock configuration directly to the clock consumer end node:
```
	clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,
		 <&rcc STM32_SRC_HSI_KER CKPER_SEL(0)>,
		 <&rcc STM32_SRC_CKPER SPI123_SEL(4)>;
```
Though, this proposal has 2 main drawbacks:
- If board dts is poorly configured, 2 consumers of the same internal clock (per_ck here) could set conflicting configurations
- This adds complexity on the consumers driver implementation side that should be able to differentiate between all these clocks 

In the end the addition of a clock mux node device felt saner.

### Limitations and next steps:
- Only STM32 SPI driver is made aware of this. All STM32 drivers should be updated to take this new clock config into account.
- It is not possible today to `clock_off` supplementary clocks as they potentially clocks multiple peripherals (or even the core itself). So, this most likely requires the use of power-domains
- On some series using "common driver" some source clocks are not yet implemented (specifically the sources based on a specific PLL output, like PLL_P, this is left as a future work). For this series, I'm focusing on a common subset of source clocks: HSE, HIS, MSI, LSE, LSI, "main" PLL output and SYSCLK.

erwango committed:
tests/drivers/clock_control: stm32h7: Add test for devices clock cfg

Add 2 clocks tests around device clock configuration on stm32h7.
For now, 'spi1_pllq_2_d1ppre_4' test variant is failed, which
illustrates issue reported in #41650.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Add support for alt clocks

Add support for alternate clocks configuration.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7_device: Use STM32_DT_CLOCKS_FOO

Make use of STM32_DT_CLOCKS_ macros to have the test work conditionally
based on alt clock presence.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_h723zg: Enable SPI

Enable SPI on nucleo_h723zg board

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/drivers/clock_control: stm32: Add _DT_CLOCK helpers

Add a DT based macros to be used by stm32 device drivers to
populate pclken[] arrays at build time.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/spi: stm32: Use alt clock freq if available

Add support for an alternate clock. If available,
alternate clock is enabled and used to get the
device clock rate.

Fixes #41650

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/bindings/clocks: Add stm32 clock mux binding

Add STM32 clock mux binding.
Only property of a node using such compatible is to
select a clock input.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers: clock_control: Add STM32 clock multiplexer driver

Add a clock multiplexer driver.
Its only function is to select a clock input.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/arm/st: stm32h7: Add perck clock-mux node

Add perck clock-mux node as disabled by default and
the include bindings helping to select it.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32h7: Add support for CKPER clock mux

Add support for CKPER clock mux.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32h7_device: Add test for CKPER source

Add 2 scenarios to test CKPER used as a clock source.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32u5: Add a _devices test

Add a stm32u5_devices test which aims at testing devices
clock control configuration on stm32u5 targets

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/dt-bindings/clock: Add stm32u5 clock bindings

This new binding allows to work on providing stm32u5 specific
alternate and complementary device clocks.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32u5: Add support for optional clocks config

This change updates stm32u5 driver to support configuration of
optional clocks on peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32u5: Fix pll_msis_80 test config

PLL input should be between 4 and 16MHz, so when MSI is set to 4MHz
fix PLLM can't be higher than 1.
Fix PLL1-NQR in consequence.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/drivers: clock_control.h: Add configure() to api

Add configure() function to clock_control API.
This function allows caller to configure a given clock.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/dt-bindings/clock: stm32: Add clock sources bindings

Add clock sources bindings on F0/F3/G0/G4/L0/L1/L4/WB/WL series.
Due to inconsistencies, some common bindings are now split:
F1 -> F0/F1/F3
L4 -> L4/G4/WB
Update .dtsi files when required

In a first step, allowed sources are limited to already supported
clocks: LSI/LSE/HSI/HSE/MSI/PCLK/PLLCLK/SYSCLK
Support for other clocks such as HSI48, SAIXCLK, ... is left for a
next step.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
include/dt-bindings/clock: stm32: Factorize Clock source binding accessors

Rename and factorize clock source bindings accessors by moving them
in common header file stm32_clock_control and remove them from
include/dt-bindings/clock/stm32XY_clock.h files

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32_common: Remove unused definitions

STM32WL_DUAL_CORE and RCC_CALC_MSI_RUN_FREQ are not used anymore.
Clean up those definitions

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32_common: Move to stm32_common_core

Move stm32_common tests to stm32_common_core before adding new folder
for device source selection tests.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers/clock_control: stm32: Add stm32_common_devices tests

Add a test section to enable device clock source selection testing.
Test targets I2C1 device which supports clock source selection
on all SOCs using this driver except L1
Initial test done on wb target.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32_common: Implement clock source selection

Similarly to what was done on U5 and H7 clock_control drivers, enable
device clock source selection.
This is done by:
-providing implementation for clock_control_configure().
-updating clock_control_get_rate() to support various possible clock
sources (SYSCLK, PLLCLK, LSE, LSI, HSI, HSE).
-providing enable_clock() to verify requested clock source exists and
is enabled.
-adding LSI and LSE device tree based initialization to
set_up_fixed_clock_sources().

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
CI reported issues are not related:
```
/__w/zephyr/zephyr/tests/drivers/spi/spi_loopback/src/spi.c: In function 'cs_ctrl_gpio_config':
/__w/zephyr/zephyr/tests/drivers/spi/spi_loopback/src/spi.c:109:2: error: 'gpio_dev' is deprecated [-Werror=deprecated-declarations]
  109 |  spi_cs.gpio_dev = device_get_binding(CS_CTRL_GPIO_DRV_NAME);
      |  ^~~~~~
```

erwango commented:
@benediktibk, @r2r0, @abxy, @lmajewski, @GeorgeCGV, @ycsin, @heinwessels, @nfaurant

lmajewski commented:
I'm just wondering if this approach to configure clocks in a generic way for STM32 MCU series is somewhat standard approach for other SoCs from this manufacturer?

In other words - is it only one off for STM32 or can it be applied to other M3/M4 ARM CPUs from other manufacturers (like e.g. NXP) ?

erwango commented:
> I'm just wondering if this approach to configure clocks in a generic way for STM32 MCU series is somewhat standard approach for other SoCs from this manufacturer?

The change is based on a new clock_control API which is generic and proposed in standalone [here](https://github.com/zephyrproject-rtos/zephyr/pull/43790). I guess the feature itself is also generic as alternate clock source is also something common across vendors (with some exceptions).
This being said, the binding proposal is STM32 specific. It also differs from STM32 Linux bindings (which also differs from STM32 uboot bindings). Aim is to make a first step in direction of a more flexible clock framework as the lack of this feature is blocking in some cases (cf https://github.com/zephyrproject-rtos/zephyr/issues/41650 but also wake-up pin selection) and has to be implemented today per driver (cf https://github.com/zephyrproject-rtos/zephyr/pull/44988), which has limitations.

erwango committed:
dts/bindings/clocks: stm32: 'clock-names' optional for source clock setting

Since implementation of clock source selection in consumer device drivers
could be achieved without usage of a clock-names property and no
example of usage is provided up to now, remove this property from existing
examples.
Additionally, make it clear in stm32 clock control binding that it is
driver's responsibility to correctly access clock source information
and use it as required.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
EDIT: I'm finding some test issues, it might not work on some cases.

GeorgeCGV commented on include/zephyr/drivers/clock_control.h:
```diff
@@ -92,13 +92,18 @@ typedef int (*clock_control_set)(const struct device *dev,
 				 clock_control_subsys_rate_t rate);
 
+typedef int (*clock_control_configure_fn)(const struct device *dev,
+				       clock_control_subsys_t sys,
```
Minor, incorrectly formatted.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32f0_clock.h:
```diff
@@ -0,0 +1,75 @@
+	 (((mask) & STM32_CLOCK_MASK_MASK) << STM32_CLOCK_MASK_SHIFT) |		\
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CFGRx register offset (RM0008.pdf) */
```
I guess it shall be `RM0091.pdf` here and below.
The `RM0008.pdf` (doesn't contain `CFGR3` register) and `RM0399.pdf` is for H7.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32f3_clock.h:
```diff
@@ -0,0 +1,85 @@
+#define TIM15_SEL(val)		STM32_CLOCK(val, 1, 10, CFGR3_REG)
+#define TIM16_SEL(val)		STM32_CLOCK(val, 1, 11, CFGR3_REG)
+#define TIM17_SEL(val)		STM32_CLOCK(val, 1, 13, CFGR3_REG)
+#define TIM20_SEL(val)		STM32_CLOCK(val, 1, 5, CFGR3_REG)
```
The shift shall be `15` instead of `5` (based on `RM0316.pdf`).

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32f3_clock.h:
```diff
@@ -0,0 +1,85 @@
+	 (((mask) & STM32_CLOCK_MASK_MASK) << STM32_CLOCK_MASK_SHIFT) |		\
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CFGRx register offset */
```
Could be great to add the RM document in the comment like `(RM0316.pdf)`.

erwango replied:
Well, I actually removed them to simplify things. One should be able to find them if needed (I do advise to use STM32McuFinder for this purpose).

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32g0_clock.h:
```diff
@@ -17,4 +15,79 @@
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CSR register offset */
+#define CCIPR_REG		0x88
```
Missing `CCIPR2_REG` or?
Based on `RM0444.pdf` and `RM0454.pdf` the `CCIPR_REG` is `0x54` and `CCIPR2_REG` is `0x58`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32g0_clock.h:
```diff
@@ -17,4 +15,79 @@
+#define USART1_SEL(val)		STM32_CLOCK(val, 3, 0, CCIPR_REG)
+#define USART2_SEL(val)		STM32_CLOCK(val, 3, 2, CCIPR_REG)
+#define USART3_SEL(val)		STM32_CLOCK(val, 3, 4, CCIPR_REG)
+#define CEC_SEL(val)		STM32_CLOCK(val, 1, 5, CCIPR_REG)
```
Based on `RM0444.pdf`  the shift shall be `6` instead of `5`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32g4_clock.h:
```diff
@@ -0,0 +1,95 @@
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CSR register offset */
+#define CCIPR_REG		0x88
```
Based on `RM0440.pdf` the address is `0x98`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32g4_clock.h:
```diff
@@ -0,0 +1,95 @@
+
+/** @brief RCC_CSR register offset */
+#define CCIPR_REG		0x88
+#define CCIPR_REG		0x9C
```
Shall be `CCIPR_REG2` instead of `CCIPR_REG`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l0_clock.h:
```diff
@@ -17,4 +15,62 @@
+#define I2C1_SEL(val)		STM32_CLOCK(val, 3, 12, CCIPR_REG)
+#define I2C3_SEL(val)		STM32_CLOCK(val, 3, 16, CCIPR_REG)
+#define LPTIM1_SEL(val)		STM32_CLOCK(val, 3, 18, CCIPR_REG)
+#define HSI48_SEL(val)		STM32_CLOCK(val, 3, 26, CCIPR_REG)
```
Based on `RM0376.pdf` the mask shall be `1`.

> `Bit26`:
> 
> - `0`: PLL USB clock selected as HSI48 clock
> - `1`: RC48 clock selected as HSI48 clock

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l1_clock.h:
```diff
@@ -6,14 +6,57 @@
+/** @brief RCC_CSR register offset */
+#define CSR_REG		0x34
+
+#define RTC_SEL(val)		STM32_CLOCK(val, 3, 16, CCIPR_REG)
```
The register shall be `CSR_REG` instead of `CCIPR_REG`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define USART2_SEL(val)		STM32_CLOCK(val, 3, 2, CCIPR_REG)
+#define USART3_SEL(val)		STM32_CLOCK(val, 3, 4, CCIPR_REG)
+#define USART4_SEL(val)		STM32_CLOCK(val, 3, 6, CCIPR_REG)
+#define LPUART1_SEL(val)	STM32_CLOCK(val, 3, 8, CCIPR_REG)
```
Based on `RM0351.pdf`, that shall be `USART5_SEL` not `LPUART1_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define USART3_SEL(val)		STM32_CLOCK(val, 3, 4, CCIPR_REG)
+#define USART4_SEL(val)		STM32_CLOCK(val, 3, 6, CCIPR_REG)
+#define LPUART1_SEL(val)	STM32_CLOCK(val, 3, 8, CCIPR_REG)
+#define I2C1_SEL(val)		STM32_CLOCK(val, 3, 10, CCIPR_REG)
```
Based on `RM0351.pdf`, that shall be `LPUART1_SEL` not `I2C2_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define USART4_SEL(val)		STM32_CLOCK(val, 3, 6, CCIPR_REG)
+#define LPUART1_SEL(val)	STM32_CLOCK(val, 3, 8, CCIPR_REG)
+#define I2C1_SEL(val)		STM32_CLOCK(val, 3, 10, CCIPR_REG)
+#define I2C2_SEL(val)		STM32_CLOCK(val, 3, 12, CCIPR_REG)
```
Based on `RM0351.pdf`, that shall be `I2C1_SEL` not `I2C2_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define LPUART1_SEL(val)	STM32_CLOCK(val, 3, 8, CCIPR_REG)
+#define I2C1_SEL(val)		STM32_CLOCK(val, 3, 10, CCIPR_REG)
+#define I2C2_SEL(val)		STM32_CLOCK(val, 3, 12, CCIPR_REG)
+#define I2C3_SEL(val)		STM32_CLOCK(val, 3, 14, CCIPR_REG)
```
Based on `RM0351.pdf`, that shall be `I2C2_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define I2C1_SEL(val)		STM32_CLOCK(val, 3, 10, CCIPR_REG)
+#define I2C2_SEL(val)		STM32_CLOCK(val, 3, 12, CCIPR_REG)
+#define I2C3_SEL(val)		STM32_CLOCK(val, 3, 14, CCIPR_REG)
+#define LPTIM1_SEL(val)		STM32_CLOCK(val, 3, 16, CCIPR_REG)
```
Based on `RM0351.pdf`, that shall be `I2C3_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define I2C2_SEL(val)		STM32_CLOCK(val, 3, 12, CCIPR_REG)
+#define I2C3_SEL(val)		STM32_CLOCK(val, 3, 14, CCIPR_REG)
+#define LPTIM1_SEL(val)		STM32_CLOCK(val, 3, 16, CCIPR_REG)
+#define LPTIM2_SEL(val)		STM32_CLOCK(val, 3, 18, CCIPR_REG)
```
Based on `RM0351.pdf`, that shall be `LPTIM1_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define I2C3_SEL(val)		STM32_CLOCK(val, 3, 14, CCIPR_REG)
+#define LPTIM1_SEL(val)		STM32_CLOCK(val, 3, 16, CCIPR_REG)
+#define LPTIM2_SEL(val)		STM32_CLOCK(val, 3, 18, CCIPR_REG)
+#define LPTIM3_SEL(val)		STM32_CLOCK(val, 3, 20, CCIPR_REG)
```
Based on `RM0351.pdf`, that shall be `LPTIM2_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define SWPMI1_SEL(val)		STM32_CLOCK(val, 1, 30, CCIPR_REG)
+#define DFSDM1_SEL(val)		STM32_CLOCK(val, 1, 31, CCIPR_REG)
+/** CCIPR2 devices */
+#define I2C4_SEL(val)		STM32_CLOCK(val, 3, 0, CCIPR_REG2)
```
`RM0351.pdf` and `RM0394.pdf` contain information only about `I2C4_SEL`, every other bit of `CCIPR_REG2` is reserved.

In what document did you find the `DFSDM_SEL`, `ADFSDM_SEL`, `DSI_SEL`, `SDMMC_SEL`, and `OSPI_SEL` in `CCIPR_REG2` of the L4?

erwango replied:
RM0432

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32u5_clock.h:
```diff
@@ -21,4 +50,73 @@
+#define OCTOSPI_SEL(val)	STM32_CLOCK(val, 3, 20, CCIPR2_REG)
+/** CCIPR3 devices */
+
 #endif /* ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32U5_CLOCK_H_ */
```
Shall be moved to the end of the file, otherwise selectors of `CCIPR3_REG` are redefined.

GeorgeCGV replied:
Shall be moved to the end of the file, otherwise selectors of CCIPR3_REG are redefined.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32wb_clock.h:
```diff
@@ -0,0 +1,91 @@
+#define CCIPR_REG		0x88
+
+/** @brief Device clk sources selection helpers */
+/** CCIPR devices */
```
Based on `RM0434.pdf`:

- There is no `USART2SEL`, `USART3SEL`, `USART4SEL`, `I2C2_SEL`, `LPTIM3_SEL`, and `SAI2_SEL`.
- `LPUART1_SEL` shift shall be `10`.
- `I2C1_SEL`  shift shall be `12`.
- `I2C3_SEL`  shift shall be `16`.
- `LPTIM1_SEL`  shift shall be `18`, `LPTIM2_SEL`  `20`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32wb_clock.h:
```diff
@@ -0,0 +1,91 @@
+	 (((mask) & STM32_CLOCK_MASK_MASK) << STM32_CLOCK_MASK_SHIFT) |		\
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CSR register offset */
```
Comment references `RCC_CSR` but related to `RCC_CCIPR`.

The `RCC_CSR` has `RFWKP_SEL` that might be added later on.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32wl_clock.h:
```diff
@@ -20,4 +18,68 @@
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CSR register offset */
+#define CCIPR_REG		0x54
```
Based on `RM0453.pdf` and `RM0461.pdf` the address is `0x88`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32wl_clock.h:
```diff
@@ -20,4 +18,68 @@
+#define LPTIM1_SEL(val)		STM32_CLOCK(val, 3, 18, CCIPR_REG)
+#define LPTIM2_SEL(val)		STM32_CLOCK(val, 3, 20, CCIPR_REG)
+#define LPTIM3_SEL(val)		STM32_CLOCK(val, 3, 22, CCIPR_REG)
+#define RNG_SEL(val)		STM32_CLOCK(val, 3, 26, CCIPR_REG)
```
Shall be `ADC_SEL` instead of `RNG_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32wl_clock.h:
```diff
@@ -20,4 +18,68 @@
+#define LPTIM2_SEL(val)		STM32_CLOCK(val, 3, 20, CCIPR_REG)
+#define LPTIM3_SEL(val)		STM32_CLOCK(val, 3, 22, CCIPR_REG)
+#define RNG_SEL(val)		STM32_CLOCK(val, 3, 26, CCIPR_REG)
+#define ADC_SEL(val)		STM32_CLOCK(val, 3, 30, CCIPR_REG)
```
Shall be `RNG_SEL` instead of `ADC_SEL`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+#define CCIPR_REG2		0x9C
+
+/** @brief Device clk sources selection helpers */
+/** CCIPR devices */
```
I suggest to go over the `REG_CCIPR` in `RM0351.pdf` once more.
Many defines are incorrect.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32g4_clock.h:
```diff
@@ -0,0 +1,95 @@
+	 (((mask) & STM32_CLOCK_MASK_MASK) << STM32_CLOCK_MASK_SHIFT) |		\
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CSR register offset */
```
Comment references `RCC_CSR` but relates to `RCC_CCIPR`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32l4_clock.h:
```diff
@@ -19,4 +17,86 @@
+	 (((mask) & STM32_CLOCK_MASK_MASK) << STM32_CLOCK_MASK_SHIFT) |		\
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CSR register offset */
```
Comment references `RCC_CSR` but relates to `RCC_CCIPR`.

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32wl_clock.h:
```diff
@@ -20,4 +18,68 @@
+	 (((mask) & STM32_CLOCK_MASK_MASK) << STM32_CLOCK_MASK_SHIFT) |		\
+	 (((val) & STM32_CLOCK_VAL_MASK) << STM32_CLOCK_VAL_SHIFT))
+
+/** @brief RCC_CSR register offset */
```
Comment references `RCC_CSR` but relates to `RCC_CCIPR`.

GeorgeCGV commented on dts/bindings/clock/st,stm32-clock-mux.yaml:
```diff
@@ -0,0 +1,43 @@
+
+compatible: "st,stm32-clock-mux"
+
+properties:
```
Most properties are available within the `base.yaml`.
Shouldn't the bindings be based on it?

erwango replied:
I intentionally didn't use base.yaml in order to limit the number of properties, as this kind of device has only one  purpose which is to set a clk source.

GeorgeCGV replied:
Makes sense. 

I don't know if there is any policy regarding bindings and using shared inclusions. Shared properties need to be kept in sync. That increases the error possibility in the future.

erwango replied:
This is right. Fixed

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -861,11 +869,20 @@ static int spi_stm32_init(const struct device *dev)
+		if (clock_control_configure(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
+					    (clock_control_subsys_t) &cfg->pclken[1],
+					    NULL) != 0) {
+			LOG_ERR("Could not select SPI soutce clock");
```
> `LOG_ERR("Could not select SPI soutce clock");`

Typo in `soutce`, shall be `source`.

GeorgeCGV commented on drivers/clock_control/clock_stm32g4.c:
```diff
@@ -25,37 +24,69 @@
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
+	if (pll_source < 0) {
```
`get_pll_source()` returns `uint32_t`.

Therefore, `if (pll_source < 0)` is always `false`.

GeorgeCGV commented on drivers/clock_control/clock_stm32g4.c:
```diff
@@ -25,37 +24,69 @@
+{
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
```
`get_pll_source()` returns `uint32_t`.
Therefore, `if (pll_source < 0)` is always `false`.

GeorgeCGV commented on drivers/clock_control/clock_stm32g0.c:
```diff
@@ -27,31 +26,64 @@
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
+	if (pll_source < 0) {
```
`get_pll_source()` returns `uint32_t`.
Therefore, `if (pll_source < 0)` is always `false`.

GeorgeCGV commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -25,30 +24,61 @@
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
+	if (pll_source < 0) {
```
`get_pll_source()` returns `uint32_t`.
Therefore, `if (pll_source < 0)` is always `false`.

GeorgeCGV commented on drivers/clock_control/clock_stm32f0_f3.c:
```diff
@@ -79,6 +79,67 @@ int config_pll_sysclock(void)
+	} else if (IS_ENABLED(STM32_PLL_SRC_HSI)) {
+		pll_input_freq = STM32_HSI_FREQ;
+	} else {
+		return -ENOTSUP;
```
it is not correct to return `-ENOTSUP` as return type is `uint32_t`.

GeorgeCGV commented on drivers/clock_control/clock_stm32f0_f3.c:
```diff
@@ -79,6 +79,67 @@ int config_pll_sysclock(void)
+	} else if (IS_ENABLED(STM32_PLL_SRC_HSI)) {
+		pll_input_freq = STM32_HSI_FREQ / 2;
+	}  else {
+		return -ENOTSUP;
```
it is not correct to return `-ENOTSUP` as return type is `uint32_t`.

GeorgeCGV commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -25,30 +24,61 @@
+		return LL_RCC_PLLSOURCE_HSE;
+	}
+
+	return -ENOTSUP;
```
it is not correct to return `-ENOTSUP` as return type is `uint32_t`.

GeorgeCGV commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -25,30 +24,61 @@
-	} else if (IS_ENABLED(STM32_PLL_SRC_HSE)) {
-		pll_source = LL_RCC_PLLSOURCE_HSE;
-	} else {
+	pll_source = get_pll_source();
```
`get_pll_source()` returns `uint32_t`.
Therefore, `if (pll_source < 0)` is always `false`.

GeorgeCGV commented on drivers/clock_control/clock_stm32g0.c:
```diff
@@ -27,31 +26,64 @@
+		return LL_RCC_PLLSOURCE_HSE;
+	}
+
+	return -ENOTSUP;
```
it is not correct to return `-ENOTSUP` as return type is `uint32_t`.

GeorgeCGV commented on drivers/clock_control/clock_stm32g4.c:
```diff
@@ -25,37 +24,69 @@
+		return LL_RCC_PLLSOURCE_HSE;
+	}
+
+	return -ENOTSUP;
```
it is not correct to return `-ENOTSUP` as return type is `uint32_t`.

Maybe `#ifdef` and `#error` combination is better?

GeorgeCGV commented on drivers/clock_control/clock_stm32l0_l1.c:
```diff
@@ -26,29 +25,59 @@
+		return LL_RCC_PLLSOURCE_HSE;
+	}
+
+	return -ENOTSUP;
```
it is not correct to return -ENOTSUP as return type is uint32_t.
Maybe #ifdef and #error combination is better?

erwango replied:
This is indeed one part that requires more thinking.
Switched to asserts

GeorgeCGV commented on drivers/clock_control/clock_stm32l0_l1.c:
```diff
@@ -26,29 +25,59 @@
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
+	if (pll_source < 0) {
```
`get_pll_source()` returns `uint32_t`.
Therefore, `if (pll_source < 0)` is always `false`.

GeorgeCGV commented on drivers/clock_control/clock_stm32l4_l5_wb_wl.c:
```diff
@@ -26,12 +25,35 @@
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
+	if (pll_source < 0) {
```
`get_pll_source()` returns `uint32_t`.
Therefore, `if (pll_source < 0)` is always `false`.

GeorgeCGV commented on drivers/clock_control/clock_stm32l4_l5_wb_wl.c:
```diff
@@ -26,12 +25,35 @@
+		return LL_RCC_PLLSOURCE_MSI;
+	}
+
+	return -ENOTSUP;
```
it is not correct to return `-ENOTSUP` as return type is `uint32_t`.
Maybe `#ifdef` and `#error` combination is better?

GeorgeCGV commented on drivers/clock_control/clock_stm32l4_l5_wb_wl.c:
```diff
@@ -40,27 +62,35 @@ int config_pll_sysclock(void)
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
+	if (pll_source < 0) {
```
`get_pll_source()` returns `uint32_t`.
Therefore, `if (pll_source < 0)` is always `false`.

erwango commented:
@GeorgeCGV Thanks for the review. I hope I fixed most of the comments. I also took care to fix tests, so it should now be fully functional (at least close to).

GeorgeCGV commented on drivers/clock_control/clock_stm32g0.c:
```diff
@@ -27,31 +26,56 @@
+	uint32_t pll_source;
+
+	pll_source = get_pll_source();
+	if (pll_source < 0) {
```
`if (pll_source < 0)` is always `false`.
Better to use `if (pll_source == 0)`.

Also, L4 and L5 define `LL_RCC_PLLSOURCE_NONE` that is `0`.

erwango replied:
Fixed

GeorgeCGV commented on include/zephyr/dt-bindings/clock/stm32g4_clock.h:
```diff
@@ -0,0 +1,95 @@
+#define ADC34_SEL(val)		STM32_CLOCK(val, 3, 30, CCIPR_REG)
+/** CCIPR2 devices */
+#define I2C4_SEL(val)		STM32_CLOCK(val, 3, 0, CCIPR_REG2)
+#define QSPI_SEL(val)		STM32_CLOCK(val, 3, 20, CCIPR_REG2)
```
```
#define I2C4_SEL(val)		STM32_CLOCK(val, 3, 0, CCIPR_REG2)
#define QSPI_SEL(val)		STM32_CLOCK(val, 3, 20, CCIPR_REG2)
```

Shall use `CCIPR2_REG` defined above.

erwango replied:
Txs. I wonder how this doesn't trigger compilation issue...
There should be something I'm missing.

GeorgeCGV commented:
> @GeorgeCGV Thanks for the review. I hope I fixed most of the comments. I also took care to fix tests, so it should now be fully functional (at least close to).

The `stm32wl_clock.h` has some critical issues: cl https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r859509619.

erwango commented:
> The `stm32wl_clock.h` has some critical issues: [#45053 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r859509619).

Fixed

GeorgeCGV commented on drivers/clock_control/clock_stm32_ll_u5.c:
```diff
@@ -71,6 +71,97 @@ static uint32_t get_startup_frequency(void)
+#elif defined(STM32_SYSCLK_SRC_HSI)) {
+	return STM32_HSI_FREQ;
+#else
+	__ASSERT(0, "No SYSCLK Source configured");
```
Missing `return 0;`

GeorgeCGV commented on drivers/clock_control/clock_stm32_mux.c:
```diff
@@ -0,0 +1,48 @@
+{
+	const struct stm32_clk_mux_config *cfg = dev->config;
+
+	if (clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
```
The formatting of arguments is a bit off.

GeorgeCGV commented on drivers/clock_control/clock_stm32_mux.c:
```diff
@@ -0,0 +1,48 @@
+};									\
+									\
+DEVICE_DT_INST_DEFINE(id, &stm32_clk_mux_init, NULL,			\
+		    NULL, &stm32_clk_mux_cfg_##id,			\
```
The formatting is a bit off.

GeorgeCGV commented on drivers/clock_control/clock_stm32l0_l1.c:
```diff
@@ -26,29 +25,44 @@
 
-	LL_RCC_PLL_ConfigDomain_SYS(pll_source, pll_mul, pll_div);
-
+	__ASSERT(0, "Invalid source");
```
Minor.

Some `__ASSERT(0, "Invalid source");` use uppercase for `Invalid`, while others (i.e.  `clock_stm32g4.c` line `40`) use lowercase `invalid source`.

Would be better to use single format approach.

erwango replied:
Done

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -489,10 +489,18 @@ static int spi_stm32_configure(const struct device *dev,
 #endif
 }
 
```
Formatting is a bit off at lines:
- `494`
- `500`
- `872`

erwango replied:
Suggestion?

GeorgeCGV replied:
494:
Added <kbd>Tab</kbd> and 2x <kbd>Space</kbd>.
```c
		if (clock_control_get_rate(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
					  (clock_control_subsys_t) &cfg->pclken[1], &clock) < 0) {
			LOG_ERR("Failed call clock_control_get_rate(pclk[1])");
			return -EIO;
		}
```

500:
Added <kbd>Tab</kbd> and 2x <kbd>Space</kbd>.
```c
		if (clock_control_get_rate(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
					  (clock_control_subsys_t) &cfg->pclken[0], &clock) < 0) {
			LOG_ERR("Failed call clock_control_get_rate(pclk[0])");
			return -EIO;
		}
```

872:
Removed 3x <kbd>Space</kbd>.
```c
	if (clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
			    (clock_control_subsys_t) &cfg->pclken[0]) != 0) {
		LOG_ERR("Could not enable SPI clock");
		return -EIO;
	}
```

erwango replied:
Done

GeorgeCGV reviewed (commented):
Open Questions

- https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r859475801

Minor

- Not corrected, but marked as fixed: https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r859564584
- https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r860631293
- https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r859427966
- https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r860619407
- https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r860620003
- https://github.com/zephyrproject-rtos/zephyr/pull/45053#discussion_r860635835

*edited: fixed links*

erwango commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -26,29 +25,44 @@
 
-	LL_RCC_PLL_ConfigDomain_SYS(pll_source, pll_m, pll_n, pll_p);
+	__ASSERT(0, "invalid source");
+}
```
Done

GeorgeCGV commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -106,6 +169,38 @@ static inline int stm32_clock_control_off(const struct device *dev,
+
+	return 0;
+#else
+	/* No srce clock available: Not supported */
```
> srce

minor, typo

GeorgeCGV commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -106,6 +169,38 @@ static inline int stm32_clock_control_off(const struct device *dev,
+	ARG_UNUSED(data);
+
+	if (enabled_clock(pclken->bus) < 0) {
+		/* Attemp to configure a src clock not available or not valid */
```
> Attemp

Minor, missing `t`.

erwango commented on drivers/clock_control/clock_stm32_ll_u5.c:
```diff
@@ -115,6 +207,33 @@ static inline int stm32_clock_control_off(const struct device *dev,
+	ARG_UNUSED(data);
+
+	if (enabled_clock(pclken->bus) < 0) {
+		/* Attemp to configure a src clock not available or not valid */
```
Fixed, thanks for the review

microbuilder reviewed (dismissed):
Can't comment on every detail of specific ST clock config being unfamiliar with some of the devices, and have to rely on CI somewhat given the size of the PR, but the general approach of specifying multiple clocks in DT seems sane to me:

```
clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00001000>,
	<&rcc STM32_SRC_PLL3_P SPI123_SEL(2)>;
```

gmarull commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -321,6 +322,66 @@ static uint32_t get_vco_output_range(uint32_t vco_input_range)
+		if (!IS_ENABLED(STM32_LSE_ENABLED)) {
+			r = -ENOTSUP;
+		}
+		break;
```
suggestion here, maybe this would be more compact

```c
	if ((src_clk == STM32_SRC_SYSCLK) ||
	    ((src_clk == STM32_SRC_HSE) && IS_ENABLED(STM32_HSE_ENABLED)) ||
	    ((src_clk == STM32_SRC_LSE) && IS_ENABLED(STM32_LSE_ENABLED)) ||
	    ...
	    ((src_clk == STM32_SRC_PLL3_R) && IS_ENABLED(STM32_PLL3_R_ENABLED))){
		return 0;
	}

	return -ENOTSUP;
```

erwango replied:
Indeed. Done. I've converted U5 as well. 
On `_common.c` there are intermixed `#ifdef FOO` that don't show up well in the `if` format. Maybe that's a matter of taste, but anyway I've kept as is.

gmarull commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -321,6 +322,66 @@ static uint32_t get_vco_output_range(uint32_t vco_input_range)
 
 #endif /* ! CONFIG_CPU_CORTEX_M4 */
 
+/** @Brief Verifies clock is part of actve clock configuration */
```
`@brief`

gmarull commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -373,6 +434,38 @@ static inline int stm32_clock_control_off(const struct device *dev,
+
+	if (enabled_clock(pclken->bus) < 0) {
+		/* Attemp to configure a src clock not available or not valid */
+		return -ENOTSUP;
```
better propagate error code

gmarull commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -861,11 +869,20 @@ static int spi_stm32_init(const struct device *dev)
+					    (clock_control_subsys_t) &cfg->pclken[1],
+					    NULL) != 0) {
+			LOG_ERR("Could not select SPI source clock");
+			return -EIO;
```
nit: propagate error code here

gmarull commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -971,17 +988,20 @@ static void spi_stm32_irq_config_func_##id(const struct device *dev)		\
+					       STM32_DT_INST_CLOCKS(id);\
+									\
 static const struct spi_stm32_config spi_stm32_cfg_##id = {		\
 	.spi = (SPI_TypeDef *) DT_INST_REG_ADDR(id),			\
-	.pclken = {							\
-		.enr = DT_INST_CLOCKS_CELL(id, bits),			\
-		.bus = DT_INST_CLOCKS_CELL(id, bus)			\
-	},								\
+	.pclken = pclken_##id,						\
+	.pclk_len = DT_INST_NUM_CLOCKS(id),				\
```
one suggestion here:

similar to what we do in pinctrl, I'd create another struct that contains the array and size, so that drivers can simply do:

```
STM32_DT_INST_CLOCKS_DEFINE(id);
```

Then device stores a reference to the defined structure.

erwango replied:
Indeed. I though about it but wanted to see how the whole feature is used across drivers before that.
Would it be fine to deploy in a second step. I agree it makes the transition a little bit 'harder' tough.

gmarull reviewed (commented):
in general lgtm, a few minor suggestions

erwango committed:
tests/drivers/clock_control: stm32: Migrate includes to <zephyr/...>

Follow up of what was done in main branch during this development.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Rebased to take into account migration to <zephyr/...> and fix related conflicts.

erwango commented:
@gmarull Thanks for the review. I've taken into account the comments and pushed once more. Agreed on the pinctrl like macro, but I prefer to introduce it in a second step.

gmarull reviewed (approved):
I haven't gone into implementation details, but the overall approach looks reasonable to me. Improvements can be made later, e.g. clocks definition.
